raco/rules.py

```diff
@@ -345,24 +345,35 @@ def descend_tree(op, cond):
             return op
         elif isinstance(op, algebra.CompositeBinaryOperator):
             # Joins and cross-products; consider conversion to an equijoin
-            left_len = len(op.left.scheme())
-            accessed = accessed_columns(cond)
-            in_left = [col < left_len for col in accessed]
-            if all(in_left):
-                # Push the select into the left sub-tree.
-                op.left = PushSelects.descend_tree(op.left, cond)
-                return op
-            elif not any(in_left):
-                # Push into right subtree; rebase column indexes
-                expression.rebase_expr(cond, left_len)
-                op.right = PushSelects.descend_tree(op.right, cond)
-                return op
-            else:
-                # Selection includes both children; attempt to create an
-                # equijoin condition
-                cols = PushSelects.is_column_equality_comparison(cond)
-                if cols:
-                    return op.add_equijoin_condition(cols[0], cols[1])
+            # Expressions containing random do not commute across joins
+            has_random = False
+            queue = [cond]
+            while queue:
+                child = queue.pop(0)
+                if isinstance(child, expression.function.RANDOM):
+                    has_random = True
+                    break
+                for c in child.get_children():
+                    queue.append(c)
```

Comment:
Looks like you're reinventing `expression.walk()` here: https://github.com/uwescience/raco/blob/master/raco/expression/expression.py#L73

I think you can replace all of L348--L365 with something similar to

```python
has_random = any(isinstance(e, RANDOM) for e in cond.walk()
```